File: App/Resources/AlgorithmDetails/flashsort/flashsort.c
#include <stdio.h>
#include <stdlib.h>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23,
                 90, 69, 51, 81, 68, 83, 32, 56};

void printList(int arr[], int n) {
  for (int i = 0; i < n; i++) {
    if (i == 0) {
      printf("[%d, ", arr[i]);
    } else if (i != n - 1) {
      printf("%d, ", arr[i]);
    } else {
      printf("%d]", arr[i]);
    }
  }
}

int classify(int value, int minValue, double c) {
  return (int)((value - minValue) * c) + 1;
}

void flashSort(int arr[], int n) {
  if (n == 0) return;

  int m = (int)(0.2 * n) + 2;

  int minValue = arr[0];
  int maxValue = arr[0];
  int maxIndex = 0;

  int i = 1;
  while (i < n - 1) {
    int small, big, bigIndex;
    if (arr[i] < arr[i + 1]) {
      small = arr[i]; big = arr[i + 1]; bigIndex = i + 1;
    } else {
      big = arr[i]; bigIndex = i; small = arr[i + 1];
    }
    if (big > maxValue) { maxValue = big; maxIndex = bigIndex; }
    if (small < minValue) { minValue = small; }
    i += 2;
  }

  int last = arr[n - 1];
  if (last < minValue) {
    minValue = last;
  } else if (last > maxValue) {
    maxValue = last;
    maxIndex = n - 1;
  }

  if (maxValue == minValue) return;

  int *L = calloc(m + 1, sizeof(int));
  double c = (m - 1.0) / (maxValue - minValue);

  for (int h = 0; h < n; h++) {
    int k = classify(arr[h], minValue, c);
    L[k] += 1;
  }

  for (int k = 2; k <= m; k++) {
    L[k] += L[k - 1];
  }

  int tmpSwap = arr[maxIndex];
  arr[maxIndex] = arr[0];
  arr[0] = tmpSwap;

  int j = 0;
  int k = m;
  int numMoves = 0;
  while (numMoves < n) {
    while (j >= L[k]) {
      j++;
      k = classify(arr[j], minValue, c);
    }

    int evicted = arr[j];
    while (j < L[k]) {
      k = classify(evicted, minValue, c);
      int location = L[k] - 1;
      int temp = arr[location];
      arr[location] = evicted;
      evicted = temp;
      L[k] -= 1;
      numMoves++;
    }
  }

  for (int idx = 1; idx < n; idx++) {
    int current = arr[idx];
    int pos = idx - 1;
    while (pos >= 0 && arr[pos] > current) {
      arr[pos + 1] = arr[pos];
      pos--;
    }
    arr[pos + 1] = current;
  }

  free(L);
}

void sort(int arr[], int n) {
  flashSort(arr, n);
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}

File: App/Resources/AlgorithmDetails/flashsort/flashsort.rb
def classify(value, min_value, c)
  ((value - min_value) * c).to_i + 1
end

def flash_sort(array)
  n = array.length
  return if n == 0

  m = (0.2 * n).to_i + 2

  min_value = array[0]
  max_value = array[0]
  max_index = 0

  i = 1
  while i < n - 1
    if array[i] < array[i + 1]
      small, big, big_index = array[i], array[i + 1], i + 1
    else
      big, big_index, small = array[i], i, array[i + 1]
    end
    if big > max_value
      max_value = big
      max_index = big_index
    end
    min_value = small if small < min_value
    i += 2
  end

  last = array[n - 1]
  if last < min_value
    min_value = last
  elsif last > max_value
    max_value = last
    max_index = n - 1
  end

  return if max_value == min_value

  l = Array.new(m + 1, 0)
  c = (m - 1.0) / (max_value - min_value)

  (0...n).each do |h|
    k = classify(array[h], min_value, c)
    l[k] += 1
  end

  (2..m).each do |k|
    l[k] += l[k - 1]
  end

  array[max_index], array[0] = array[0], array[max_index]

  j = 0
  k = m
  num_moves = 0
  while num_moves < n
    while j >= l[k]
      j += 1
      k = classify(array[j], min_value, c)
    end

    evicted = array[j]
    while j < l[k]
      k = classify(evicted, min_value, c)
      location = l[k] - 1
      temp = array[location]
      array[location] = evicted
      evicted = temp
      l[k] -= 1
      num_moves += 1
    end
  end

  (1...n).each do |idx|
    current = array[idx]
    pos = idx - 1
    while pos >= 0 && array[pos] > current
      array[pos + 1] = array[pos]
      pos -= 1
    end
    array[pos + 1] = current
  end
end

def sort(array)
  flash_sort(array)
  array
end

array = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(array)
p array
